crates/datasources/src/cassandra/mod.rs

```diff
@@ -0,0 +1,167 @@
+mod builder;
+mod errors;
+mod exec;
+use async_stream::stream;
+use async_trait::async_trait;
+use datafusion::arrow::array::{
+    ArrayRef, Float32Builder, Float64Builder, Int16Builder, Int32Builder, Int8Builder,
+    StringBuilder, TimestampMillisecondBuilder,
+};
+use datafusion::arrow::datatypes::{
+    DataType, Field, Fields, Schema as ArrowSchema, SchemaRef as ArrowSchemaRef, TimeUnit,
+};
+
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::datasource::TableProvider;
+use datafusion::error::{DataFusionError, Result as DatafusionResult};
+use datafusion::execution::context::SessionState;
+use datafusion::execution::TaskContext;
+use datafusion::logical_expr::{Expr, TableProviderFilterPushDown, TableType};
+use datafusion::physical_expr::PhysicalSortExpr;
+use datafusion::physical_plan::metrics::{ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning, RecordBatchStream,
+    SendableRecordBatchStream, Statistics,
+};
+use datafusion_ext::metrics::DataSourceMetricsStreamAdapter;
+pub use errors::*;
+use futures::Stream;
+use futures::StreamExt;
+use scylla::frame::response::result::{ColumnType, CqlValue, Row};
+
+use scylla::transport::session::Session;
+use scylla::SessionBuilder;
+use std::any::Any;
+
+use std::fmt;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use self::exec::CassandraExec;
+
+struct CassandraAccess {
+    session: Session,
+}
+
+fn try_convert_dtype(ty: &ColumnType) -> Result<DataType> {
+    Ok(match ty {
+        ColumnType::Ascii => DataType::Utf8,
+        ColumnType::Date => DataType::Date64,
+        ColumnType::Double => DataType::Float64,
+        ColumnType::Duration => DataType::Duration(TimeUnit::Nanosecond),
+        ColumnType::Float => DataType::Float32,
+        ColumnType::Int => DataType::Int32,
+        ColumnType::Text => DataType::Utf8,
+        ColumnType::Timestamp => DataType::Timestamp(TimeUnit::Millisecond, None),
+        ColumnType::SmallInt => DataType::Int16,
+        ColumnType::TinyInt => DataType::Int8,
+        ColumnType::Uuid => DataType::Utf8,
+        ColumnType::BigInt => DataType::Int64,
+        ColumnType::List(inner) | ColumnType::Set(inner) => {
+            let inner = try_convert_dtype(inner)?;
+            DataType::new_list(inner, true)
+        }
+        _ => return Err(CassandraError::UnsupportedDataType(format!("{:?}", ty))),
+    })
+}
+
+impl CassandraAccess {
+    pub async fn try_new(conn_str: impl AsRef<str>) -> Result<Self> {
+        let session = SessionBuilder::new().known_node(conn_str).build().await?;
+        Ok(Self { session })
+    }
+    async fn get_schema(&self, table: &str) -> Result<ArrowSchema> {
+        let query = format!("SELECT * FROM {table} LIMIT 1");
```

Comment:
this is what every other sql datasource is doing. I c/p'd this from clickhouse.

Reply:
This is something we should open an issue for to check. I initially developed the data sources thinking we wouldn't need to escape since I assumed parsing/validation would fail higher up, and so the value being passed here has gone through _some_ validation. But I'm actually unsure what happens for double quoted identifiers...

Reply:
The values do go through the sqlparser before we get them in the table funcs. it's not foolproof, but sqlparser does provide us a layer of protection against sql injection. 

so something like this will fail during parsing
```sql
select * from read_cassandra('127.0.0.1:9000' ';DROP TABLE test.supported_dtypes; --');
```

Reply:
> so the value being passed here has gone through some validation.

it goes through sqlparser, but i don't think they are designed to protect against injection. We'll likely want to handle this kind of sanitizing in `FuncParamValue`

Reply:
> > so the value being passed here has gone through some validation.
> 
> it goes through sqlparser, but i don't think they are designed to protect against injection. We'll likely want to handle this kind of sanitizing in `FuncParamValue`

We may need even more than that. For example with external databases, what would happens with something like `SELECT * FROM my_external_db.public."1; select 2"`? Would that get the `select 2` query through?

Reply:
Opened https://github.com/GlareDB/glaredb/issues/2350